src/smriprep/cli/run.py

```diff
@@ -495,6 +528,32 @@ def build_workflow(opts, retval):
     bids_dir = opts.bids_dir.resolve()
     layout = BIDSLayout(str(bids_dir), validate=False)
     subject_list = collect_participants(layout, participant_label=opts.participant_label)
+    session_list = opts.session_label or []
+
+    subject_session_list = []
+    for subject in subject_list:
+        sessions = layout.get_sessions(
+            scope='raw',
+            subject=subject,
+            session=session_list or Query.OPTIONAL,
+            suffix=['T1w', 'T2w'],  # TODO: Track supported modalities globally
+        )
+        if not sessions:
+            if opts.subject_anatomical_reference == 'sessionwise':
+                logger.warning(
+                    '--subject-anatomical-reference "sessionwise" was requested, but no sessions '
+                    f'were found for subject {subject}.'
+                )
+            subject_session_list.append((subject, None))
+            continue
+
+        if opts.subject_anatomical_reference == 'sessionwise':
+            for session in sessions:
+                subject_session_list.append((subject, session))
+        else:
+            # This will use all sessions either found by layout or passed in via --session-id
+            # MG: Should session names be concatenated into a label to preserve provenance?
+            subject_session_list.append((subject, sessions))
 
     bids_filters = json.loads(opts.bids_filter_file.read_text()) if opts.bids_filter_file else None
```

Comment:
I'm concerned this may further filter sessions and change the assumptions done thus far.

Reply:
I agree there could be unseen interactions when collecting data, but not sure the best way to account for this without greatly increasing code complexity - we still want to preserve using `--bids-filter-file` in tandem with this session pairing. Open to suggestions